kbcstorage/base.py

```diff
@@ -0,0 +1,139 @@
+"""
+Base classes for constructing the client.
+
+Primarily exposes a base Endpoint class which deduplicates functionality across
+various endpoints, such as tables, workspaces, jobs, etc. as described in the
+`Storage API documentation`.
+
+
+.. _Storage API documentation:
+    http://docs.keboola.apiary.io/
+"""
+from urllib.parse import urljoin
+
+import requests
+
+
+class Endpoint:
+    """
+    Base class for implementing a single endpoint related to a single entities
+    as described in the Storage API.
+
+    Attributes:
+        path (str): URL for this endpoint.
+        token (str): A key for the Storage API.
+    """
+    def __init__(self, root, extension, token):
+        """
+        Create an endpoint.
+
+        Args
+            root (str): Root url of API. eg.
+                "https://connection.keboola.com/v2/storage/"
+            extension (str): Extension of url for the endpoint. eg. "buckets"
+            token (str): A key for the Storage API. Can be found in the storage
+            console.
+        """
+        self.path = urljoin(root, extension)
+        self.token = token
+
+    def _get(self, params=[], extra_headers={}):
```

Comment:
I see the `Endpoint`s `_get()` `_post`,... methods merely as a very thin wrappers for the `requests` native http methods, which only appends the `X-StorageApi-token: TOKEN` header to the request. I think it makes sense to delegate the url construction to the endpoints. And if some repetitive code appears, we can always refactor it into the base `Endpoint` class. 

fwiw it's probably not even necessary to hide them with the leading `_underscore` but make them "public" i.e `Endpoint._get(*args, **kwargs)` --> `Endpoint.get(*args, **kwargs)`.